File: 25. Reverse Nodes in k-Group Hard/solution.ts
class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val: number, next: ListNode | null = null) {
        this.val = val;
        this.next = next;
    }
}

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    function reverseLinkedList(head: ListNode | null): ListNode | null {
        let prev: ListNode | null = null;
        let current: ListNode | null = head;
        while (current !== null) {
            let nextNode: ListNode | null = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }
        return prev;
    }

    function countNodes(node: ListNode | null): number {
        let count: number = 0;
        while (node !== null) {
            count++;
            node = node.next;
        }
        return count;
    }

    let dummy: ListNode = new ListNode(0);
    dummy.next = head;
    let prevGroupEnd: ListNode = dummy;

    let nodesCount: number = countNodes(head);

    while (nodesCount >= k) {
        let groupStart: ListNode | null = prevGroupEnd.next;
        let current: ListNode | null = groupStart;
        for (let i = 0; i < (k - 1); i++) {
            current = current!.next;
        }

        let groupEnd: ListNode | null = current;
        let nextGroupStart: ListNode | null = groupEnd!.next;

        groupEnd!.next = null;
        prevGroupEnd.next = reverseLinkedList(groupStart);
        groupStart.next = nextGroupStart;

        prevGroupEnd = groupStart;
        nodesCount -= k;
    }

    return dummy.next;
}

let head1: ListNode = new ListNode(1);
head1.next = new ListNode(2);
head1.next.next = new ListNode(3);
head1.next.next.next = new ListNode(4);
head1.next.next.next.next = new ListNode(5);

let k1: number = 2;
console.log(`Input: [1, 2, 3, 4, 5], k = ${k1}`);
let result1: ListNode | null = reverseKGroup(head1, k1);
while (result1 !== null) {
    console.log(result1.val);
    result1 = result1.next;
}


let head2: ListNode = new ListNode(1);
head2.next = new ListNode(2);
head2.next.next = new ListNode(3);
head2.next.next.next = new ListNode(4);
head2.next.next.next.next = new ListNode(5);

let k2: number = 3;
console.log(`Input: [1, 2, 3, 4, 5], k = ${k2}`);
let result2: ListNode | null = reverseKGroup(head2, k2);
while (result2 !== null) {
    console.log(result2.val);
    result2 = result2.next;
}

File: 25. Reverse Nodes in k-Group Hard/solution.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	head1 := &ListNode{Val: 1}
	head1.Next = &ListNode{Val: 2}
	head1.Next.Next = &ListNode{Val: 3}
	head1.Next.Next.Next = &ListNode{Val: 4}
	head1.Next.Next.Next.Next = &ListNode{Val: 5}

	k1 := 2
	fmt.Printf("Input: [1,2,3,4,5], k = %d\n", k1)
	result1 := reverseKGroup(head1, k1)
	for result1 != nil {
		fmt.Printf("%d ", result1.Val)
		result1 = result1.Next
	}
	fmt.Println()

	head2 := &ListNode{Val: 1}
	head2.Next = &ListNode{Val: 2}
	head2.Next.Next = &ListNode{Val: 3}
	head2.Next.Next.Next = &ListNode{Val: 4}
	head2.Next.Next.Next.Next = &ListNode{Val: 5}

	k2 := 3
	fmt.Printf("Input: [1,2,3,4,5], k = %d\n", k2)
	result2 := reverseKGroup(head2, k2)
	for result2 != nil {
		fmt.Printf("%d ", result2.Val)
		result2 = result2.Next
	}
	fmt.Println()
}

func reverseLinkedList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	current := head
	for current != nil {
		nextNode := current.Next
		current.Next = prev
		prev = current
		current = nextNode
	}
	return prev
}

func countNodes(node *ListNode) int {
	count := 0
	for node != nil {
		count++
		node = node.Next
	}
	return count
}

func reverseKGroup(head *ListNode, k int) *ListNode {
	dummy := &ListNode{Val: 0, Next: head}
	prevGroupEnd := dummy

	nodesCount := countNodes(head)

	for nodesCount >= k {
		groupStart := prevGroupEnd.Next
		current := groupStart
		for i := 0; i < k-1; i++ {
			current = current.Next
		}

		groupEnd := current
		nextGroupStart := groupEnd.Next

		groupEnd.Next = nil
		prevGroupEnd.Next = reverseLinkedList(groupStart)
		groupStart.Next = nextGroupStart

		prevGroupEnd = groupStart
		nodesCount -= k
	}

	return dummy.Next
}
